### PR TITLE
feat: Add javascript definitions to .sqlx files

### PR DESCRIPTION
**Pull request by HampB**
## Description
This draft PR introduces a new feature that allows users to navigate directly to the row of a variable or function when clicked. The functionality is demonstrated below:

Definition Peek Example

The implementation is still a work in progress and may be somewhat inefficient or messy at this stage. I’d appreciate any feedback, change requests, or suggestions for improvement. Feel free to use this as a foundation to restructure or refine as needed.

## Outstanding Tasks
- **Improve Variable Validation**: Refine logic to better identify actual JavaScript variables (currently, it captures anything starting or ending with a period).
- **Enhance Search Logic**: Update the search logic to accurately retrieve the correct row number, especially when the same name appears multiple times in the `.js` file.
  
- **Consolidate Classes**: Consider merging the current three classes into a single `DataformDefinition` class to streamline the logic for retrieving the appropriate definition. This consolidation isn’t critical for now, but it may improve maintainability and scalability in future iterations.

## Questions / Next Steps
I'm considering how best to structure functions for reusability, especially if we decide to add onHover functionality. Any guidance or suggestions on this would be greatly appreciated!

**TODOs**

- [x] Update documentation before merge

### CODE REVIEW

**Comment by ashish10alex** [comment]
Hi @HampB ,
thanks for the PR. The basic functionality is definitely there albeit needs refining. Would be doing some tests and suggesting changes as as we continue to work on it. 

----

One of the edge case I found was the following 

```js
config {
  type: 'table'
}

js {
  const SIX_MONTHS_AGO_FROM_NOW = parameters.getDateMonthsBefore(new Date(), -6)
}

select ${SIX_MONTHS_AGO_FROM_NOW} as some_time
```

would have been nice if the Dataform graph had some metadata of where the javascript variables are coming from. I feel our best bet for the above is use the output of `getMetadataForSqlxFileBlocks`  function and check if there is a js block and then find for the declaration there first ?

**Comment by HampB** [comment]
Done some fixes to adress issues in https://github.com/ashish10alex/vscode-dataform-tools/pull/52 

Still unfixed issues:
- Cant click the module part of a statement like this ${module.varOrFunction} (should take you to the top of that file?)
- The search in .js doc matches the first occurrence of text (pass the searchTerm with a prexif of const, var, function, and so on or do a regex based search?) 

The search quite often takes you to the wrong place in our production repo.

**Comment by ashish10alex** [comment]
> The search in .js doc matches the first occurrence of text (pass the searchTerm with a prexif of const, var, function, and so on or do a regex based search?)

yes, agree, ill write some code to make this change

**Comment by ashish10alex** [comment]
putting the file in here which I am using to test the features 

```js
config {
  type: 'table'
}

/*
create a parameters.js file in the includes directory and add 
getDateMonthsBefore function

function getDateMonthsBefore(date, months) {
  date.setMonth(date.getMonth() + months);
  let date_string = date.toISOString().split('T')[0];
  date_string = `'${date_string}'`;
  return date_string

}
*/

js {
  let some_mod = require("includes/parameters.js")
  const six_months_ago_from_now = some_mod.getDateMonthsBefore(new Date(), -6)
  const four_months_ago_from_now = some_mod.getDateMonthsBefore(new Date(), -6)
  const some_month = "january"
  const some_number = 11;
  let some_coniditon = true;
  let another_coniditon = true;
  var some_floating_points = 9.4;
}

select 
  *
, ${six_months_ago_from_now}  some_time
, ${four_months_ago_from_now}  some_time
, ${some_number}  some_number
, ${some_coniditon}  some_condition
, ${another_coniditon}  another_coniditon
, ${some_floating_points}  some_floating_points
, "${some_month}"  some_month 
, ${some_mod.getDateMonthsBefore(new Date(), -4)} date_wt_func
, ${parameters.OXRC_DATA_STARTING_POINT}  as date_from_file

FROM ${ref("0100_CALENDAR")}
```

File: src/definitionProvider.ts [hunk 1]
```diff
@@ -145,4 +142,142 @@ export class DataformRequireDefinitionProvider implements vscode.DefinitionProvi
...
+            return undefined;
+        }
+        const line = document.lineAt(position.line).text;
+        // const start = wordRange.start.character;
+        // const end = wordRange.end.character;
+
+        // This will often not be inside ${}, rather it will be inside js {}
+        if (line.includes('require("')){
+            return findModuleDefinition(document, workspaceFolder, searchTerm);
+        }
```
[the remark this answers is not in the document]
Comment (HampB) [reply src/definitionProvider.ts:254]: You're probably right, my thougt was that it's unlikely that it will be something else since the line contains "require(".

A check that only return if the value is not undefined is probably good enough?

[the remark this answers is not in the document]
Comment (ashish10alex) [reply src/definitionProvider.ts:254]: > A check that only return if the value is not undefined is probably good enough?

yes, that would suffice, thanks

[the remark this answers is not in the document]
Comment (HampB) [reply src/definitionProvider.ts]: fixed

**Comment by HampB** [comment]
Thanks for the additions @ashish10alex!

I think this should be production ready soon.

But first another issue that has come up in one of our iterations is that, in some cases, all words on the line receive the function or variable definition instead of just the intended variable or function name.

For example, this code assigns the function definition to `"Var1"`, `"Var2"`, `AS`, and `unique_key`:

```javascript
js {
  let utils = require("includes/functions/utils.js");
}

SELECT
${utils.createUniqueKey(["Var1", "Var2"])} AS unique_key
```

The same thing seems to happen when the identifier name is the same as the variable inside of the ${}. Like this:
`${some_coniditon} AS some_condition`

One quick solution could involve adding a check like this inside the `DataformJsDefinitionProvider`:

```javascript
const startIndex = line.indexOf(variableOrFunctionName);
const endIndex = startIndex + variableOrFunctionName.length;

if (position.character < startIndex || position.character > endIndex) {
    return undefined;
}
```

I can see if i can find a better solution and give it a try later, if you do not have any clear way of solving it? :-)

**Comment by ashish10alex** [comment]
Hi @HampB , thanks, I was unable to reproduce the issue you were having. I used the following code to test. Please let me know what I am doing differently 

```js
config {
  type: 'table'
}

/*
let someVariable = "createUniqueKey";

function createUniqueKey(someList){
  return someList.map(item => item.toString()).join("")
}

module.exports = {
    createUniqueKey,
    someVariable,
};
*/

js {
  let some_mod = require("includes/functions/utils.js")
}


select 
  "${some_mod.someVariable}" as someVariable
, "${some_mod.createUniqueKey(['Var1', 'Var2'])}" AS unique_key
```

https://github.com/user-attachments/assets/8a2377ae-0315-423b-a4c8-91d0e3e0ff28

**Comment by ashish10alex** [comment]
thanks @HampB for the clarification. Please go ahead with the fix that you were suggesting. Will review that

**Comment by HampB** [comment]
This PR should be ready for final review and testing.
I've been running this branch during the day and have not ran into any issues so far.

**Comment by ashish10alex** [comment]
Thanks @HampB , I will update the docs sometime this weekend and merge this PR
